hazelcast/src/hazelcast/client/compact.cpp

```diff
@@ -0,0 +1,678 @@
+{
+    std::vector<field_descriptor*> fixed_size_fields;
+    std::vector<field_descriptor*> boolean_fields;
+    std::vector<field_descriptor*> variable_size_fields;
+
+    std::map<std::string, field_descriptor> sorted_fields(
+      field_definition_map_.begin(), field_definition_map_.end());
+    for (auto& item : sorted_fields) {
+        field_descriptor& descriptor = item.second;
+        field_kind kind = descriptor.field_kind;
+        if (field_operations::get(kind).kind_size_in_byte_func() ==
+            field_kind_based_operations::VARIABLE_SIZE) {
+            variable_size_fields.push_back(&descriptor);
+        } else if (kind == field_kind::BOOLEAN) {
+            boolean_fields.push_back(&descriptor);
+        } else {
+            fixed_size_fields.push_back(&descriptor);
+        }
+    }
+
+    std::sort(
+      fixed_size_fields.begin(), fixed_size_fields.end(), kind_size_comparator);
+
+    int offset = 0;
+    for (auto descriptor : fixed_size_fields) {
+        descriptor->offset = offset;
+        offset += field_operations::get(descriptor->field_kind)
+                    .kind_size_in_byte_func();
+    }
+
+    int bit_offset = 0;
+    for (auto descriptor : boolean_fields) {
+        descriptor->offset = offset;
+        descriptor->bit_offset = bit_offset % util::Bits::BITS_IN_BYTE;
+        bit_offset++;
+        if (bit_offset % util::Bits::BITS_IN_BYTE == 0) {
+            offset += 1;
+        }
+    }
+    if (bit_offset % util::Bits::BITS_IN_BYTE != 0) {
+        offset += 1;
+    }
+
+    fixed_size_fields_length_ = offset;
+
+    int index = 0;
+    for (auto descriptor : variable_size_fields) {
+        descriptor->index = index;
+        index++;
+    }
+
+    for (auto& item : sorted_fields) {
+        field_definition_map_[item.first] = item.second;
+    }
+
+    number_of_var_size_fields_ = index;
+    schema_id_ = rabin_finger_print::fingerprint64(type_name_, sorted_fields);
+}
+
+int64_t
+schema::schema_id() const
+{
+    return schema_id_;
+}
+
+size_t
+schema::number_of_var_size_fields() const
+{
+    return number_of_var_size_fields_;
+}
+
+size_t
+schema::fixed_size_fields_length() const
+{
+    return fixed_size_fields_length_;
+}
+
+const std::string&
+schema::type_name() const
+{
+    return type_name_;
+}
+
+const std::unordered_map<std::string, field_descriptor>&
+schema::fields() const
+{
+    return field_definition_map_;
+}
+
+std::ostream&
+operator<<(std::ostream& os, const schema& schema)
+{
+    os << "type name " << schema.type_name() << ",number of var size fields "
+       << schema.number_of_var_size_fields() << ",fixed size fields length "
+       << schema.fixed_size_fields_length() << ",fields {";
+    for (const auto& item : schema.fields()) {
+        os << item.first << " " << item.second << ",";
+    }
+    os << "}";
+    return os;
+}
+
+std::ostream&
+operator<<(std::ostream& os, const field_descriptor& field_descriptor)
+{
+    os << field_descriptor.field_kind;
```

Comment:
Will this print a number? How useful will that be to the user? Can we print string form ? Or maybe the user will not see this output at all, I am not sure.

Reply:
In cpp, without keeping another set of array, map it is not possible to print the text. 
See https://stackoverflow.com/questions/3342726/c-print-out-enum-value-as-text

Not sure if it will worth the effort.